modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/TransportRankEvalAction.java

```diff
@@ -118,11 +122,28 @@ protected void doExecute(RankEvalRequest request, ActionListener<RankEvalRespons
             types = ratedRequest.getTypes().toArray(types);
             templatedRequest.types(types);
 
-            client.search(templatedRequest, searchListener);
+            RequestTask task = new RequestTask(templatedRequest, searchListener);
+            taskQueue.add(task);
+        }
+
+        // Execute top n tasks, further execution is triggered in RankEvalActionListener
+        for (int i = 0; (i < qualityTask.getMaxConcurrentSearches() && (! taskQueue.isEmpty())); i++) {
```

Comment:
Maybe its possible to simplify termination condition here, something like Math.min(ratedRequests.size(), qualityTask.getMaxConcurrentSearches())? I found the and/not statement a bit difficult to follow at first glance.